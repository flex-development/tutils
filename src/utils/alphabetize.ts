/**
 * @file Utilities - alphabetize
 * @module tutils/utils/alphabetize
 */

import { SortOrder, type CompareResult } from '#src/enums'
import type { Fn, Nilable, Writable } from '#src/types'
import type AlphabetizeOptions from './alphabetize.options'
import sort from './sort'

/**
 * Sorts an array alphabetically without modifying it.
 *
 * A `mapper` can be used to convert array items to strings.
 *
 * @see {@linkcode AlphabetizeOptions}
 *
 * @todo examples
 *
 * @template T - Array to sort
 *
 * @param {T} arr - Array to sort
 * @param {Fn<[T[number]], string>} mapper - Array item interpolator
 * @param {Nilable<AlphabetizeOptions>} [options={}] - Comparsion options
 * @return {Writable<T>} Alphabetically sorted copy of target array
 */
const alphabetize = <T extends readonly unknown[]>(
  arr: T,
  mapper: Fn<[T[number]], string>,
  options: Nilable<AlphabetizeOptions> = null
): Writable<T> => {
  const {
    caseFirst = 'upper',
    locales,
    numeric = true,
    order = SortOrder.ASC,
    ...opts
  } = (options ??= {})

  return sort(arr, (a: T[number], b: T[number]): CompareResult => {
    /**
     * Target string.
     *
     * @const {string} target
     */
    const target: string = mapper(order === SortOrder.ASC ? a : b)

    /**
     * String to compare against {@linkcode target}.
     *
     * @const {string} that
     */
    const that: string = mapper(order === SortOrder.ASC ? b : a)

    return target.localeCompare(that, locales, {
      ...opts,
      caseFirst,
      numeric,
      usage: 'sort'
    })
  })
}

export default alphabetize
