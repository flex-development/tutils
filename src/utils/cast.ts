/**
 * @file Utilities - cast
 * @module tutils/utils/cast
 */

/**
 * Casts `value`.
 *
 * @template T - Return value type
 *
 * @param {unknown} value - Value to cast
 * @return {T} `value`
 */
const cast = <T>(value: unknown): T => value as T

export default cast
