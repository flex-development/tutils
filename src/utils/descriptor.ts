/**
 * @file Utilities - descriptor
 * @module tutils/utils/descriptor
 */

import type { PropertyDescriptor } from '#src/interfaces'
import type { PropertyKey } from '#src/types'

/**
 * Returns an object describing the configuration of a specific property.
 *
 * @see https://developer.mozilla.org/docs/Web/JavaScript/Reference/Global_Objects/Object/getOwnPropertyDescriptor
 *
 * @todo examples
 *
 * @template T - Property value type
 *
 * @param {unknown} target - Value to look for `key` in
 * @param {PropertyKey} key - Property key
 * @return {PropertyDescriptor<T>} Property descriptor object
 */
const descriptor = <T = any>(
  target: unknown,
  key: PropertyKey
): PropertyDescriptor<T> => Object.getOwnPropertyDescriptor(target, key) ?? {}

export default descriptor
