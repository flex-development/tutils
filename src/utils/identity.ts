/**
 * @file Utilities - identity
 * @module tutils/utils/identity
 */

/**
 * Returns `value` without modifications.
 *
 * @template T - Value type
 *
 * @param {T} value - Value to return
 * @return {T} `value`
 */
function identity<T>(value: T): T {
  return value
}

export default identity
