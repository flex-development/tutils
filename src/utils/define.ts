/**
 * @file Utilities - define
 * @module tutils/utils/define
 */

import type { PropertyDescriptor } from '#src/interfaces'
import type { PropertyKey } from '#src/types'
import cast from './cast'
import description from './descriptor'
import hasOwn from './has-own'

/**
 * Adds a property to an object, or modifies attributes of an existing property.
 *
 * @see {@linkcode PropertyDescriptor}
 *
 * @todo examples
 *
 * @template T - Object on which to add or modify property
 * @template V - Property value
 * @template U - Updated object type
 *
 * @param {T} obj - Object on which to add or modify the property
 * @param {PropertyKey} property - Property to add or modify
 * @param {PropertyDescriptor<V>} [descriptor] - Property descriptor object
 * @return {U} Updated object
 */
const define = <T extends object, V = any, U = T>(
  obj: T,
  property: PropertyKey,
  descriptor: PropertyDescriptor<V> = {}
): U => {
  return Object.defineProperty(cast(obj), property, {
    ...(hasOwn(obj, property)
      ? description(obj, property)
      : hasOwn(descriptor, 'get') || hasOwn(descriptor, 'set')
      ? { configurable: true, enumerable: true }
      : { configurable: true, enumerable: true, writable: true }),
    ...descriptor
  })
}

export default define
