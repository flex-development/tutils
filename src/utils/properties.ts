/**
 * @file Utilities - properties
 * @module tutils/utils/properties
 */

import type { Nilable, PropertyKey, Stringify } from '#src/types'
import cast from './cast'
import enumerable from './enumerable'
import type PropertiesOptions from './properties.options'
import select from './select'

/**
 * Construct a union of own property keys.
 *
 * @template K - Property key type
 */
type Properties<K extends PropertyKey> = K extends symbol ? K : Stringify<K>

/**
 * Returns an array containing own properties of a given target value.
 *
 * The own properties of an object are those that are defined directly on that
 * object, and are not inherited from the object's prototype.
 *
 * Non-object arguments are [coerced to objects][1]. Enumerable own properties
 * of a string are its indices.
 *
 * [1]: https://developer.mozilla.org/docs/Web/JavaScript/Reference/Global_Objects/Object#object_coercion
 *
 * @see {@linkcode Properties}
 * @see {@linkcode PropertiesOptions}
 *
 * @todo examples
 *
 * @template T - Target value
 * @template K - Property key type
 *
 * @param {T} target - Target value
 * @param {Nilable<PropertiesOptions>} [options] - Property retrieval options
 * @param {Nilable<boolean>} [options.enu] - Enumerable properties only
 * @return {Properties<K>[]} Own properties of `target`
 */
const properties = <T, K extends PropertyKey>(
  target: T,
  options?: Nilable<PropertiesOptions>
): Properties<K>[] => {
  return select(
    cast<Properties<K>[]>([
      ...Object.getOwnPropertySymbols(target),
      ...Object.getOwnPropertyNames(target)
    ]),
    property => (options?.enu ? enumerable(target, property) : true)
  )
}

export { properties as default, type Properties }
