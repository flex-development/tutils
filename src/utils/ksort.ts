/**
 * @file Utilities - ksort
 * @module tutils/utils/ksort
 */

import type { PropertyDescriptor } from '#src/interfaces'
import type { Nilable, ObjectCurly, Objectify } from '#src/types'
import alphabetize from './alphabetize'
import cast from './cast'
import define from './define'
import descriptor from './descriptor'
import identity from './identity'
import isArray from './is-array'
import isObjectPlain from './is-object-plain'
import type KsortOptions from './ksort.options'
import properties from './properties'

/**
 * Returns an object with sorted keys.
 *
 * The initial target object **will** be modified.
 *
 * @see {@linkcode KsortOptions}
 *
 * @todo examples
 *
 * @template T - Object containing keys to sort
 *
 * @param {T} obj - Object containing keys to sort
 * @param {Nilable<KsortOptions>} [options] - Key sorting options
 * @return {T} Target object with sorted keys
 */
const ksort = <T extends Objectify<any>>(
  obj: T,
  options?: Nilable<KsortOptions>
): T => {
  // do nothing if obj is not an array or plain object
  if (!isArray(obj) && !isObjectPlain(obj)) return obj

  /**
   * Sorted values cache.
   *
   * @const {WeakMap<object, object>} cache
   */
  const cache: WeakMap<object, object> = new WeakMap()

  /**
   * Sorts the keys of plain objects in the given array.
   *
   * @template T - Array containing plain objects
   *
   * @param {T} arr - Array containing plain objects
   * @return {T} New array containing plain objects with sorted keys
   */
  const arrsort = <T extends unknown[]>(arr: T): T => {
    // return cached output value
    if (cache.has(arr)) return cast(cache.get(arr))

    // cache array
    cache.set(arr, arr)

    // sort keys of plain objects in arr
    for (const [key, item] of arr.entries()) {
      if (isArray(item)) arr[key] = arrsort(cast(item))
      if (isObjectPlain(item)) arr[key] = sort(item)
    }

    return arr
  }

  /**
   * Removes all properties from an object.
   *
   * @template T - Target object
   *
   * @param {T} obj - Target object
   * @return {T} Empty object
   */
  const clear = <T extends ObjectCurly>(obj: T): T => {
    for (const key of properties(obj)) Reflect.deleteProperty(obj, key)
    return obj
  }

  /**
   * Returns a new object with sorted keys.
   *
   * @template T - Object to sort
   *
   * @param {T} object - Object to sort
   * @return {T} New object with sorted keys
   */
  const sort = <T extends ObjectCurly>(object: T): T => {
    // return cached output value
    if (cache.has(object)) return cast(cache.get(object))

    /**
     * Property names mapped to property descriptors.
     *
     * @const {[string | symbol,  PropertyDescriptor][]}
     */
    const props: [string | symbol, PropertyDescriptor][] = [
      ...alphabetize(Object.getOwnPropertyNames(object), identity, options),
      ...Object.getOwnPropertySymbols(object)
    ].map(key => [key, descriptor(object, key)])

    /**
     * New object with sorted keys.
     *
     * @const {T} ret
     */
    const ret: T = clear(object)

    // cache object with sorted keys
    cache.set(object, ret)

    // define properties of new object and recursively sort keys
    for (const [key, descriptor] of props) {
      if (options?.deep) {
        switch (true) {
          case isArray(descriptor.value):
            descriptor.value = arrsort(cast<unknown[]>(descriptor.value))
            break
          case isObjectPlain(descriptor.value):
            descriptor.value = sort(cast<ObjectCurly>(descriptor.value))
            break
          default:
            break
        }
      }

      // define property
      define(ret, key, descriptor)
    }

    return ret
  }

  return cast(
    isArray(obj) ? (options?.deep ? arrsort(cast(obj)) : obj) : sort(obj)
  )
}

export default ksort
