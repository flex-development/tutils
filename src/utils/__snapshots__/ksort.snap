// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`unit:utils/ksort > arrays > deep > should return obj with item keys sorted deeply 1`] = `
[
  {
    "make": "lexus",
    "model": "lc",
    "vin": "0WBW1G4D29TC62167",
    "year": 2023,
  },
  [Circular],
]
`;

exports[`unit:utils/ksort > plain objects > deep > should return obj with nested and top-level keys sorted 1`] = `
{
  "circular": [Circular],
  "driver": {
    "first_name": "tres",
    "last_name": "koelpin",
    "nanoid": "mhR2OWSEhQFXkwU1NJY23",
  },
  "make": "lexus",
  "model": "lc",
  "riders": [
    {
      "first_name": "adrian",
      "last_name": "rodriguez",
      "uuid": "19fd6413-7fc9-4e7f-a5b5-9a3689d25134",
    },
  ],
  "vin": "0WBW1G4D29TC62167",
  "year": 2023,
}
`;

exports[`unit:utils/ksort > plain objects > should return obj with top-level keys sorted 1`] = `
{
  "make": "lexus",
  "model": "lc",
  "vin": "0WBW1G4D29TC62167",
  "year": 2023,
}
`;
